blog post: add OpenTelemetry-Apache APISIX integration blog (#1185)
open-telemetry/opentelemetry.io: content/en/blog/2022/apisix.md

---

> title: Integrating Apache APISIX and OpenTelemetry to Enhance Observability Capabilities

Author: Title is modied now.

linkTitle: Apache APISIX-Opentelemetry Integration

> date: 2022-03-03

Author: We can fix this at the end, when we know the exact date of publication.

author: Haochao Zhuang, Fei Han
---

> This article introduces you to the Apache APISIX `opentelemetry` plugin concept and how to enable and deploy the plugin.

Author: Plugin is registered according to the guidelines :)

## Background Information

> OpenTelemetry is an open source telemetry data acquisition and processing system. It not only provides various SDKS for application side telemetry data collection and reporting, but also data collection side for data receiving, processing and exporting. Export to any or more OpenTelemetry backends, such as Jaeger, Zipkin, and OpenCensus, by configuration. You can view the list of plug-ins that have adapted the OpenTelemetry Collector in the [opentelemetry collector contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib) library.

Author: Since the blog's title is changed, this becomes somewhat unnecessary to change imo.

![Architecture-Present](/img/apisix/architecture-present.png)

## Plugin Introduction

The `opentelemetry` plugin of Apache APISIX implements Tracing data collection based on OpenTelemetry native standard (OTLP/HTTP), and sends it to OpenTelemetry Collector through HTTP protocol. This feature will be supported online in Apache APISIX 2.13.0.

> Since the Agent/SDK of OpenTelemetry has nothing to do with the back-end implementation, when the application integrates the Agent/SDK of OpenTelemetry, the user can easily and freely change the observability backend service without any perception on the application side, such as switching from Zipkin into Jaeger.

Author: Clarified.

The `opentelemetry` plug-in integrates the OpenTelemetry Agent/SDK in Apache APISIX, which can collect traced requests, generate `trace` and forward them to the OpenTelemetry Collector.

The opentelemetry plugin is located on the Agent side in the above figure, but currently only supports the `trace` part, and does not support the `logs` and `metrics` protocols of OpenTelemetry.

## Enable the Plugin

You need to enable `opentelemetry` plugin and modify collector configuration in `conf/config.yaml` configuration file.

> We assume that you have already deployed the OpenTelemetry Collector and enabled the [OTLP HTTP Receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md).

Reviewer: ```suggestion
We assume that you have already deployed the OpenTelemetry Collector on the same node as the APISIX and enabled the [OTLP HTTP Receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md).
```

Author: Fixed.

> If you have not completed the deployment, you can refer to the Scenario Example section in the next section to complete the deployment of OpenTelemetry Collector.

The default port of the OTLP HTTP Receiver is `4318`, and the address of the `collector` is the HTTP Receiver address of the OpenTelemetry Collector. For related fields, please refer to the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/).

For specific configuration, please refer to the following example:

```shell
plugins
  ... # Other plugins that have been enabled
  - opentelemetry
plugin_attr:
  ...
  opentelemetry:
    trace_id_source: x-request-id
    resource:
      service.name: APISIX
    collector:
      address: 127.0.0.1:4318 # OTLP HTTP Receiver address
      request_timeout: 3
```

### Method 1: Enable the Plugin for a Specific Route

> In order to show the test effect more conveniently, `sampler` is temporarily set to full sampling in the example to ensure that `trace` data is generated after each request is traced, so that you can view `trace` related data on the Web UI. You can also set relevant parameters according to the actual situation.

Reviewer: can you add a note on how sampling affects the overhead introduced by the plugin. Any perf data you can refer to would be useful in this blog post

Author: Yeah unfortunately we don't have that type of data. Can we skip it?

Reviewer: sure, it may be a quesiton customers will have, but if we don't have answer yet, it is fine. Perhaps can be mentioned as a future work

You need to make sure you have enabled the `opentelemetry` plugin and reload Apache APISIX.

1. You can refer to the following example to create a route and enable the opentelemetry plugin for sampling:
  
  ```shell
  curl http://127.0.0.1:9080/apisix/admin/routes/1 \
  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
  -X PUT -d '
  {
      "uri": "/hello",
      "plugins": {
          "opentelemetry": {
              "sampler": {
                  "name": "always_on",
              }
          }
      },
      "upstream": {
          "type": "roundrobin",
          "nodes": {
              "127.0.0.1:7080": 1
          }
      }
  }'
  ```

2. Modify the `./examples/demo/otel-collector-config.yaml` file to add the OTLP HTTP Receiver.

   ```shell
   receivers:
    otlp:
      protocols:
        grpc:
        http:${ip:port}   # add OTLP HTTP Receiver，default port is 4318

    ```

3. Modify `docker-compose.yaml` file.

  You need to modify the configuration file, change the interface address of Client calling Server to the address of Apache APISIX, and map the ports of OTLP HTTP Receiver and Server services to local.

>   The following example is the complete docker-compose.yaml after the configuration is modified:

Reviewer: ```suggestion
  The following example is the complete `docker-compose.yaml` after the configuration is modified:
```

Author: Fixed.

        - "4317"        # OTLP gRPC receiver
        - "4318:4318"   # Add OTLP HTTP Receiver port mapping
        - "55670:55679" # zpages extension
      depends_on:
        - jaeger-all-in-one
        - zipkin-all-in-one

    demo-client:
      build:
        dockerfile: Dockerfile
        context: ./client
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=otel-collector:4317
        - DEMO_SERVER_ENDPOINT=http://172.17.0.1:9080/hello # APISIX address
      depends_on:
        - demo-server

    demo-server:
      build:
        dockerfile: Dockerfile
        context: ./server
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=otel-collector:4317
      ports:
        - "7080:7080" # Map the Server port to the host
      depends_on:
        - otel-collector

    prometheus:
      container_name: prometheus
      image: prom/prometheus:latest
      volumes:
        - ./prometheus.yaml:/etc/prometheus/prometheus.yml
      ports:
        - "9090:9090"
  ```

It should be noted that demo-client.environment.DEMO_SERVER_ENDPOINT needs to be changed to your Apache APISIX address, and ensure that it can be accessed normally in the container.

> Of course, you can also deploy Apache APISIX through docekr-compose.yaml. For details, please refer to the Apache APISIX official documentation.

Reviewer: ```suggestion
Of course, you can also deploy Apache APISIX through `docekr-compose.yaml`. For details, please refer to the Apache APISIX official documentation.
```

Author: Fixed.

> ### Step 3: Verify the Outputs

Reviewer: either in this section or next one would be very important to add the following:

- With plugin you presumably get visibility into cases when the request does not reach the demo-server. Is there some kind of policies may be implemented on APISIX? What span will be generated in this case?
- Important to note, that even if demo-server is not instrumented, by enabling this plugin you are getting visibility into demo-server behavior. So it is a poor man substitute of a real instrumentation of demo-server which is extremely easy to do and out of the box provides a lot of value.

Author: Ditto, Haochao is working on it, should be delivered by this coming Tuesday.

Author: Updated.

After the redeployment is completed, you can access the Jaeger UI or Zipkin UI to see that the Span of APISIX is included in the Trace, as shown below:

![Jaeger UI](/img/apisix/jaeger-ui-3.png)

> ![Zipkin UI](/img/apisix/zipkin-ui-2.png)

Reviewer: on this picture it would be nice to select the APISIX span so it's attributes will be shown in the right panel

Author: Haochao is working on it, it should be delivered by on this coming Tuesday.

Author: Updated.

> ## Disable the Plugin

Reviewer: please add what will change if it's disabled. Both spans (client and server) will still be connected, but APISIX span will be missing, correct?

Author: Yes. Add a sentence to clarify.

OpenTelemetry Agent/SDK and adopts its features in Apache APISIX. It can
collect traced requests, generate `trace`, and forward them to the OpenTelemetry
Collector. It supports the `trace` protocol, and it will support the `logs` and
`metrics` protocols of OpenTelemetry in the next version.

## Enable the Plugin

You need to enable `opentelemetry` plugin and modify collector configuration in
`conf/config.yaml` configuration file.

We assume that you have already deployed the OpenTelemetry Collector on the same
node as the APISIX and enabled the [OTLP HTTP Receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md).

> Need help completing deployment of the OpenTelemetry Collector?
> See the scenario [Example](#example) below.

The default port of the OTLP HTTP Receiver is `4318`, and the address of the
`collector` is the HTTP Receiver address of the OpenTelemetry Collector. For
related fields, please refer to the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/).

For specific configuration, please refer to the following example:

> ```shell

Author: Updated.

```

After the access is successful, you can see the details of the span similar to
`/put` in the Jaeger UI, and you can see that the custom tags in the route are
displayed in the Tags list: `http_x-custom-ot-key` and `route_id`.

![Jaeger UI](/img/apisix/jaeger-ui-1.png)

You need to note that the `additional_attributes` configuration is set to take
values from Apache APISIX and Nginx variables as `attribute` values, so
`additional_attributes` must be a valid Apache APISIX or Nginx variable. It also
includes HTTP Header, but when fetching http_header, you need to add `http_` as
the prefix of the variable name. If the variable does not exist, the `tag` will
not be displayed.

## Example

This scenario example deploys Collector, Jaeger, and Zipkin as backend services
by simply modifying the official example of OpenTelemetry Collector, and starts
two sample applications (Client and Server), where Server provides an HTTP
service, and Client will cyclically call the server provided by the server. HTTP
interface, resulting in a call chain consisting of two spans.

### Step 1: Deploy OpenTelemetry

The following uses docker-compose as an example. For other deployments, please

> refer to the OpenTelemetry official documentation.

Reviewer 2: Change the last sentence to something like:

> For other deployments, see ADD-A-LINK-HERE

Author: Updated.

application is shown in the following figure.

![topology of the demo](/img/apisix/demo-topology.png)

The Trace data reporting process is as follows. Among them, since Apache APISIX
is deployed separately and not in the network of docker-compose, Apache APISIX
accesses the OTLP HTTP Receiver of OpenTelemetery Collector through the locally
mapped port (`127.0.0.1:4138`).

![Trace data reporting process](/img/apisix/trace-data-flow.png)

You need to make sure you have enabled the `opentelemetry` plugin and reload
Apache APISIX.

You can refer to the following example to create a route and enable
the`opentelemetry` plugin for sampling:
  
```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
-X PUT -d '
{
  "uri": "/hello",
  "plugins": {
      "opentelemetry": {
          "sampler": {
            "name": "always_on"
          }
      }
  },
  "upstream": {
      "type": "roundrobin",
      "nodes": {
          "127.0.0.1:7080": 1
      }
  }
}'
```

Modify the `./examples/demo/otel-collector-config.yaml` file to add the OTLP
HTTP Receiver.  

```shell
  receivers:
  otlp:
    protocols:
      grpc:

>       http:${ip:port}   # add OTLP HTTP Receiver，default port is 4318

Author: Updated.

```

Modify `docker-compose.yaml` file.

You need to modify the configuration file, change the interface address of
Client calling Server to the address of Apache APISIX, and map the ports of
OTLP HTTP Receiver and Server services to local.

The following example is the complete `docker-compose.yaml` after the
configuration is modified:

  ```yaml
  version: "2"

>   services:

Reviewer 3: Seems we don't need the indent here, right ?

Author: Yeah, updated.

        - zipkin-all-in-one

    demo-client:
      build:
        dockerfile: Dockerfile
        context: ./client
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=otel-collector:4317
        - DEMO_SERVER_ENDPOINT=http://172.17.0.1:9080/hello # APISIX address
      depends_on:
        - demo-server

    demo-server:
      build:
        dockerfile: Dockerfile
        context: ./server
      environment:
        - OTEL_EXPORTER_OTLP_ENDPOINT=otel-collector:4317
      ports:
        - "7080:7080" # Map the Server port to the host
      depends_on:
        - otel-collector

    prometheus:
      container_name: prometheus
      image: prom/prometheus:latest
      volumes:
        - ./prometheus.yaml:/etc/prometheus/prometheus.yml
      ports:
        - "9090:9090"
  ```

It should be noted that `demo-client.environment.DEMO_SERVER_ENDPOINT` needs to
be changed to your Apache APISIX address, and ensure that it can be accessed
normally in the container.

Of course, you can also deploy Apache APISIX through `docekr-compose.yaml`. For
details, please refer to the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/how-to-build/#installation-via-docker).

> 

Reviewer 2: For details, please --> For details, see [Installation via Docker](...)

Author: Updated.

### Step 3: Verify the Outputs

After the redeployment is completed, you can access the Jaeger UI or Zipkin UI
to see that the Span of APISIX is included in the Trace, as shown below:

![Jaeger UI](/img/apisix/jaeger-ui-3.png)

![Zipkin UI](/img/apisix/zipkin-ui-2.png)

When demo-server is not instrumented, you can still getting visibility of the
demo-server behavior by enabling this plugin. Although this is not a typical
case, it is a poor-man substitute of a real instrumentation of demo-server and
provides a lot of value.

![upstream-not-instrumented](/img/apisix/upstream-not-instrumented.png)

When the request does not reach the demo-server, the output would not include
the span of demo-server.

![demo-server disconneted](/img/apisix/demo-server-disconnected.png)

## Disable the Plugin

If you do not need trace collection of a route temporarily, you only need to
modify the route configuration and delete the part of `opentelemetry` under
`plugins` in the configuration.

If you enabled `opentelemetry` globally by binding Global Rules, you can remove
the configuration of the `opentelemetry` global plugin.

> Please note that disabling the `opentelemetry` plugin only results in

Reviewer 2: Please note ... --> Note ...

Author: Updated.

---
title: Apache APISIX Integrates with OpenTelementry to Collect Tracing Data
linkTitle: Apache APISIX-Opentelemetry Integration

> date: 2022-03-13

Reviewer 2: Update the publication date.

Author: Changed to `date: 2022-03-25`.

author: Haochao Zhuang, Fei Han
---

This article introduces the Apache APISIX's `opentelemetry` plugin concept and
how to enable and deploy the plugin.

## Background Information

OpenTelemetry is an open source telemetry data acquisition and processing
system. It not only provides various SDKs for application side telemetry data
collection and reporting, but also provides data collection side for data
receiving, processing, and exporting. Export to any or more OpenTelemetry
backends, such as Jaeger, Zipkin, and OpenCensus. You can view the list of

> plugins that have adapted the OpenTelemetry Collector in the [OpenTelemetry Registry](https://opentelemetry.io/registry/?s=collector&component=&language=).

Author: Changed to `[OpenTelemetry Registry](/registry/?s=collector&component=&language=)`

backend services they want, such as Zipkin and Jaeger, without affectting the
application side.

The `opentelemetry` plugin is located on the Agent side. It integrates the
OpenTelemetry Agent/SDK and adopts its features in Apache APISIX. It can
collect traced requests, generate `trace`, and forward them to the OpenTelemetry
Collector. It supports the `trace` protocol, and it will support the `logs` and
`metrics` protocols of OpenTelemetry in the next version.

## Enable the Plugin

You need to enable `opentelemetry` plugin and modify collector configuration in
`conf/config.yaml` configuration file.

We assume that you have already deployed the OpenTelemetry Collector on the same
node as the APISIX and enabled the [OTLP HTTP Receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md).

> If you have not completed the deployment, you can refer to the Scenario
> Example section in the next section to complete the deployment of
> OpenTelemetry Collector.

The default port of the OTLP HTTP Receiver is `4318`, and the address of the
`collector` is the HTTP Receiver address of the OpenTelemetry Collector. For

> related fields, please refer to the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/).

Reviewer 2: Changes:
- please refer to --> see
- Drop "official" -- unless you have a strong reason to want to contrast "official" vs. any non-official documentation.

Author: Did a global search, replace "please refer to", "you can refer to", etc with "see".
Dropped "official".

> For specific configuration, please refer to the following example:

Reviewer 2: Suggestion: change "For specific ..." to:

> A typical configuration might look like this:

Author: Updated.

> ```shell

Reviewer 2: This is YAML, is it not? If so, change to code-block language to reflect this.

Author: Yes, it is YAML. Updated.

`additional_attributes`, you can refer to the following configuration:

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1001 \
  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
  -X PUT -d '
{
    "uri": "/put",
    "plugins": {
        "opentelemetry": {
            "sampler": {
                "name": "always_on"
            },
            "additional_attributes":[
                "route_id",
                "http_x-custom-ot-key"
            ]
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "httpbin.org:80": 1
        }
    }
}'
```

## Test and Verify the Plugin

You can enable `opentelemetry` plugin in any of the above three methods. The
following example uses the example of method three to create a route. After the
creation is successful, you can refer to the following commands to access the
route:

```shell
curl -X PUT -H `x-custom-ot-key: test-ot-val` http://127.0.0.1:9080/put
```

After the access is successful, you can see the details of the span similar to
`/put` in the Jaeger UI, and you can see that the custom tags in the route are
displayed in the Tags list: `http_x-custom-ot-key` and `route_id`.

> ![Jaeger UI](/img/apisix/jaeger-ui-1.png)

Author: Modified all screenshots.

> ![Jaeger UI](/img/apisix/jaeger-ui-1.png)

Reviewer 2: Great, thanks.

values from Apache APISIX and Nginx variables as `attribute` values, so
`additional_attributes` must be a valid Apache APISIX or Nginx variable. It also
includes HTTP Header, but when fetching http_header, you need to add `http_` as
the prefix of the variable name. If the variable does not exist, the `tag` will
not be displayed.

## Example

This scenario example deploys Collector, Jaeger, and Zipkin as backend services
by simply modifying the official example of OpenTelemetry Collector, and starts
two sample applications (Client and Server), where Server provides an HTTP
service, and Client will cyclically call the server provided by the server. HTTP
interface, resulting in a call chain consisting of two spans.

### Step 1: Deploy OpenTelemetry

The following uses docker-compose as an example. For other deployments, please
refer to the OpenTelemetry official documentation.

You can refer to the following command to deploy:

```shell
git clone https://github.com/open-telemetry/opentelemetry-collector-contrib.git
cd opentelemetry-collector-contrib/examples/demo
docker-compose up -d
```

Enter `http://127.0.0.1:16886` (Jaeger UI) or `http://127.0.0.1:9411/zipkin`

> (Zipkin UI) in the browser. If it can be accessed normally, the deployment is

Reviewer 2: I'd vote to keep "the deployment is successful".

> (Zipkin UI) in the browser. If it can be accessed normally, the deployment is

Reviewer 3: LGTM

> (Zipkin UI) in the browser. If it can be accessed normally, the deployment is

Author: Solved.

The Apache APISIX service is introduced, and the topology of the final
application is shown in the following figure.

![topology of the demo](/img/apisix/demo-topology.png)

The Trace data reporting process is as follows. Among them, since Apache APISIX
is deployed separately and not in the network of docker-compose, Apache APISIX
accesses the OTLP HTTP Receiver of OpenTelemetery Collector through the locally
mapped port (`127.0.0.1:4138`).

![Trace data reporting process](/img/apisix/trace-data-flow.png)

You need to make sure you have enabled the `opentelemetry` plugin and reload
Apache APISIX.

You can refer to the following example to create a route and enable
the`opentelemetry` plugin for sampling:
  
```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 \
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
-X PUT -d '
{
  "uri": "/hello",
  "plugins": {
      "opentelemetry": {
          "sampler": {
            "name": "always_on"
          }
      }
  },
  "upstream": {
      "type": "roundrobin",
      "nodes": {
          "127.0.0.1:7080": 1
      }
  }
}'
```

Modify the `./examples/demo/otel-collector-config.yaml` file to add the OTLP
HTTP Receiver.  

> ```shell

Reviewer 2: shell --> yaml

Author: Updated.

provides a lot of value.

![upstream-not-instrumented](/img/apisix/upstream-not-instrumented.png)

When the request does not reach the demo-server, the output would not include
the span of demo-server.

![demo-server disconneted](/img/apisix/demo-server-disconnected.png)

## Disable the Plugin

If you do not need trace collection of a route temporarily, you only need to
modify the route configuration and delete the part of `opentelemetry` under
`plugins` in the configuration.

If you enabled `opentelemetry` globally by binding Global Rules, you can remove
the configuration of the `opentelemetry` global plugin.

Please note that disabling the `opentelemetry` plugin only results in
disconnecting the APISIX span, the client and server spans will remain
connected.

## Summary

After Apache APISIX integrates OpenTelemetery, it can easily connect with many
Trace systems on the market. Apache APISIX is also actively cooperating with
communities to create a more powerful ecosystem.

Apache APISIX is also currently working on additional plugins to support
integration with more services, so if you're interested, feel free to start a

> discussion thread in our [GitHub Discussion](https://github.com/apache/apisix/discussions)

Reviewer 2: How about:

> feel free to [start a discussion] on GitHub or ...

[start a discussion]: https://github.com/apache/apisix/discussions

Author: Updated.

---
title: Apache APISIX Integrates with OpenTelementry to Collect Tracing Data
linkTitle: Apache APISIX-Opentelemetry Integration
date: 2022-03-13
author: Haochao Zhuang, Fei Han
---

> This article introduces the Apache APISIX's `opentelemetry` plugin conceptand

Reviewer: ```suggestion
This article introduces the Apache APISIX's `opentelemetry` plugin concept and
```

Author: Solved.

how to enable and deploy the plugin.

## Background Information

OpenTelemetry is an open source telemetry data acquisition and processing
system. It not only provides various SDKs for application side telemetry data
collection and reporting, but also provides data collection side for data

> receiving, processing and exporting. Export to any or more OpenTelemetry

Reviewer: ```suggestion
receiving, processing, and exporting. Export to any or more OpenTelemetry
```

Author: Solved.

backends, such as Jaeger, Zipkin, and OpenCensus. You can view the list of
plugins that have adapted the OpenTelemetry Collector in the

> [opentelemetry collector contrib](https://github.com/open-telemetry/opentelemetry-collector-contrib)

Reviewer: this page doesn't have the list of plugins. Maybe this link would work better: https://opentelemetry.io/registry/?s=collector&component=all&language=

Author: Solved.

library.

![Architecture-Present](/img/apisix/architecture-present.png)

## Plugin Introduction

The `opentelemetry` plugin of Apache APISIX implements Tracing data collection

> based on OpenTelemetry native standard (OTLP/HTTP), and sends it to

Reviewer: ```suggestion
and sends it to
```
I think this is not relevant and confusing piece of information

Author: Solved.

OpenTelemetry Collector through HTTP protocol. Apache APISIX starts to support
this feature in v2.13.0.

One of OpenTelemetry's special features is that the Agent/SDK of OpenTelemetry
is not locked with back-end implementation, which gives users flexibilities on
choosing their own back-end services. In other words, users can choose the
backend services they want, such as Zipkin and Jaeger, without affectting the
application side.

The `opentelemetry` plugin is located on the Agent side. It integrates the
OpenTelemetry Agent/SDK and adopts its features in Apache APISIX. It can

> collect traced requests, generate `trace` and forward them to the OpenTelemetry

Reviewer: ```suggestion
collect traced requests, generate `trace`, and forward them to the OpenTelemetry
```

Author: Solved.

`metrics` protocols of OpenTelemetry in the next version.

## Enable the Plugin

You need to enable `opentelemetry` plugin and modify collector configuration in
`conf/config.yaml` configuration file.

We assume that you have already deployed the OpenTelemetry Collector on the same
node as the APISIX and enabled the [OTLP HTTP Receiver](https://github.com/open-telemetry/opentelemetry-collector/blob/main/receiver/otlpreceiver/README.md).

> If you have not completed the deployment, you can refer to the Scenario
> Example section in the next section to complete the deployment of
> OpenTelemetry Collector.

The default port of the OTLP HTTP Receiver is `4318`, and the address of the
`collector` is the HTTP Receiver address of the OpenTelemetry Collector. For
related fields, please refer to the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/).

For specific configuration, please refer to the following example:

```shell
plugins
  ... # Other plugins that have been enabled
  - opentelemetry
plugin_attr:
  ...
  opentelemetry:
    trace_id_source: x-request-id
    resource:
      service.name: APISIX
    collector:

>       address: 127.0.0.1:4318 # OTLP HTTP Receiver address

Reviewer: out of curiosity, is default http or https? Is it configureable?

Author: It is default http now. By looking at the source code[1], I don't think https is supported. But if that is super desired feature, you can create an issue in the repo, and then we can add it to the next version.

[1]https://github.com/apache/apisix/blob/master/apisix/plugins/opentelemetry.lua#L71

      request_timeout: 3
```

### Method 1: Enable the Plugin for a Specific Route

In order to show the test effect more conveniently, `sampler` is temporarily set
to full sampling in the example to ensure that `trace` data is generated after
each request is traced, so that you can view `trace` related data on the Web UI.
You can also set relevant parameters according to the actual situation.

```shell
curl http://127.0.0.1:9080/apisix/admin/routes/1 \
  -H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
  -X PUT -d '
{
    "uri": "/get",
    "plugins": {
        "opentelemetry": {
            "sampler": {
                "name": "always_on"
            }
        }
    },
    "upstream": {
        "type": "roundrobin",
        "nodes": {
            "httpbin.org:80": 1
        }
    }
}'
```

### Method 2: Enable the Plugin Globally

You can also enable `opentelemetry` plugin through the Apache APISIX Plugins
feature. After the global configuration is complete, you still need to create
the route, otherwise it will not be possible to test.

```shell
curl 'http://127.0.0.1:9080/apisix/admin/global_rules/1' \

> -H 'X-API-KEY:  edd1c9f034335f136f87ad84b625c8f1' \

Reviewer: ```suggestion
-H 'X-API-KEY: edd1c9f034335f136f87ad84b625c8f1' \
```

Author: Solved.

-X PUT -d '{
    "plugins": {
        "opentelemetry": {
            "sampler": {
                "name": "always_on"
            }
        }
    }
}'
```

### Method 3: Customize Labels for Span through additional_attributes

For the configuration of `sampler` and `additional_attributes`, you can refer to

> the [Apache APISIX officialdocumentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/#attributes),

Reviewer: ```suggestion
the [Apache APISIX official documentation](https://apisix.apache.org/docs/apisix/next/plugins/opentelemetry/#attributes),
```

Author: Solved.
